### DISCUSSION TITLE
CPU Problem Executing SPARQL

### DISCUSSION THREAD

**Discussion by naturzukunft** [eclipse-rdf4j/rdf4j#3908] [Q&A] [answered]
I am adding models to a repository. to check the versions of the models to be added, I need to query the maximum version for an identifier for each insert.

```
PREFIX schema: <https://schema.org/> 
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#> 
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>

SELECT (MAX(?version) as ?maxVersion)
WHERE 
{
	?objectId schema:identifier "%s" .
	?objectId schema:version ?version .
	?objectId rdf:type schema:CreativeWork .
}

```    
At the beginning, determining the versions (via HTTP SPARQL!) takes ~200ms. When I get closer to 2000 models, the duration for the query increases significantly.
With 2400 models I am already at a duration of 450ms.

I was expecting more of a memory problem, but from ~ 2000 objects onwards you can clearly see that the cpu is close to 100%. Network load drops, memory remains fairly constant. From this I deduce that my server, which executes the query, is overloading the cpu. (Of course, a lot is running in parallel there)

But I have only inserted a fraction of the models. So my import will never be finished.

**Comment by hmottestad** [comment]
Can you tell us a bit more about your setup?
 - Java version
 - Amount of RAM
 - RDF4J version
 - NativeStore or MemoryStore?
 - Reasoning?
 - What isolation level are you using?


Are you able use the explanation feature for SPARQL queries so we can see what's going on? 
 - https://rdf4j.org/documentation/programming/repository/#explaining-queries
 - Preferably using both `Explanation.Level.Timed` and `Explanation.Level.Unoptimized`

**Reply by naturzukunft** [reply]
- Server & Client is running on my local maschine.
- Java version: 17
- Amount of RAM On my local machine 16GB (jvmArguments: -Xms256m -Xmx768m)
- RDF4J version_ 3.7.7
- NativeStore or MemoryStore?: NativeStore on local disk
- Reasoning?: no
- What isolation level are you using?: default

## explanation feature at the beginning with an almost empty triple store:
**TimedDistinct** 
```
(resultSizeActual=12, totalTimeActual=6.27ms, selfTimeActual=1.54ms)
   Projection (resultSizeActual=12, totalTimeActual=4.74ms, selfTimeActual=0.113ms)
   ├──ProjectionElemList
   │     ProjectionElem "identifier"
   │     ProjectionElem "version"
   └──Join (JoinIterator) (resultSizeActual=12, totalTimeActual=4.62ms, selfTimeActual=0.677ms)
      ╠══StatementPattern (costEstimate=16, resultSizeEstimate=48, resultSizeActual=12, totalTimeActual=1.67ms, selfTimeActual=1.67ms)
      ║     Var (name=subject)
      ║     Var (name=_const_fb4ba132_uri, value=https://schema.org/identifier, anonymous)
      ║     Var (name=_const_78bed9c9_lit_e2eec718_0, value="6472c8bbf6504b56b473a1e6a10fcc8d", anonymous)
      ╚══Join (JoinIterator) (resultSizeActual=12, totalTimeActual=2.28ms, selfTimeActual=0.508ms)
         ├──StatementPattern (costEstimate=1, resultSizeEstimate=468, resultSizeActual=12, totalTimeActual=0.537ms, selfTimeActual=0.537ms)
         │     Var (name=subject)
         │     Var (name=_const_f5e5585a_uri, value=http://www.w3.org/1999/02/22-rdf-syntax-ns#type, anonymous)
         │     Var (name=_const_eaa085c9_uri, value=https://schema.org/CreativeWork, anonymous)
         └──StatementPattern (costEstimate=19, resultSizeEstimate=365, resultSizeActual=12, totalTimeActual=1.23ms, selfTimeActual=1.23ms)
               Var (name=subject)
               Var (name=_const_880c9cef_uri, value=https://schema.org/version, anonymous)
               Var (name=version)
```

**UnoptimizedDistinct** 
```
(resultSizeActual=12, totalTimeActual=26.2ms, selfTimeActual=0.125ms)
   Projection (resultSizeActual=12, totalTimeActual=26.1ms, selfTimeActual=0.022ms)
   ├──ProjectionElemList
   │     ProjectionElem "identifier"
   │     ProjectionElem "version"
   └──Join (JoinIterator) (resultSizeActual=12, totalTimeActual=26.1ms, selfTimeActual=4.19ms)
      ╠══StatementPattern (costEstimate=16, resultSizeEstimate=48, resultSizeActual=12, totalTimeActual=3.14ms, selfTimeActual=3.14ms)
      ║     Var (name=subject)
      ║     Var (name=_const_fb4ba132_uri, value=https://schema.org/identifier, anonymous)
      ║     Var (name=_const_78bed9c9_lit_e2eec718_0, value="6472c8bbf6504b56b473a1e6a10fcc8d", anonymous)
      ╚══Join (JoinIterator) (resultSizeActual=12, totalTimeActual=18.7ms, selfTimeActual=3.61ms)
         ├──StatementPattern (costEstimate=1, resultSizeEstimate=7.6K, resultSizeActual=12, totalTimeActual=7.5ms, selfTimeActual=7.5ms)
         │     Var (name=subject)
         │     Var (name=_const_f5e5585a_uri, value=http://www.w3.org/1999/02/22-rdf-syntax-ns#type, anonymous)
         │     Var (name=_const_eaa085c9_uri, value=https://schema.org/CreativeWork, anonymous)
         └──StatementPattern (costEstimate=78, resultSizeEstimate=6.0K, resultSizeActual=12, totalTimeActual=7.63ms, selfTimeActual=7.63ms)
               Var (name=subject)
               Var (name=_const_880c9cef_uri, value=https://schema.org/version, anonymous)
               Var (name=version)
```

## explanation feature at a later time, where the cpu is a bit more hot ;-)
**TimedDistinct** 
```
(resultSizeActual=14, totalTimeActual=78.6ms, selfTimeActual=0.252ms)
   Projection (resultSizeActual=14, totalTimeActual=78.3ms, selfTimeActual=0.038ms)
   ├──ProjectionElemList
   │     ProjectionElem "identifier"
   │     ProjectionElem "version"
   └──Join (JoinIterator) (resultSizeActual=14, totalTimeActual=78.3ms, selfTimeActual=24.6ms)
      ╠══StatementPattern (costEstimate=19, resultSizeEstimate=56, resultSizeActual=14, totalTimeActual=16.1ms, selfTimeActual=16.1ms)
      ║     Var (name=subject)
      ║     Var (name=_const_fb4ba132_uri, value=https://schema.org/identifier, anonymous)
      ║     Var (name=_const_d5c851ae_lit_e2eec718_0, value="27bccef5a7ef49dab2414f6eb6cd8ef4", anonymous)
      ╚══Join (JoinIterator) (resultSizeActual=14, totalTimeActual=37.6ms, selfTimeActual=4.2ms)
         ├──StatementPattern (costEstimate=1, resultSizeEstimate=7.6K, resultSizeActual=14, totalTimeActual=18.9ms, selfTimeActual=18.9ms)
         │     Var (name=subject)
         │     Var (name=_const_f5e5585a_uri, value=http://www.w3.org/1999/02/22-rdf-syntax-ns#type, anonymous)
         │     Var (name=_const_eaa085c9_uri, value=https://schema.org/CreativeWork, anonymous)
         └──StatementPattern (costEstimate=78, resultSizeEstimate=6.0K, resultSizeActual=14, totalTimeActual=14.6ms, selfTimeActual=14.6ms)
               Var (name=subject)
               Var (name=_const_880c9cef_uri, value=https://schema.org/version, anonymous)
               Var (name=version)
```

**UnoptimizedDistinct** 
```
(resultSizeActual=14, totalTimeActual=78.6ms, selfTimeActual=0.252ms)
   Projection (resultSizeActual=14, totalTimeActual=78.3ms, selfTimeActual=0.038ms)
   ├──ProjectionElemList
   │     ProjectionElem "identifier"
   │     ProjectionElem "version"
   └──Join (JoinIterator) (resultSizeActual=14, totalTimeActual=78.3ms, selfTimeActual=24.6ms)
      ╠══StatementPattern (costEstimate=19, resultSizeEstimate=56, resultSizeActual=14, totalTimeActual=16.1ms, selfTimeActual=16.1ms)
      ║     Var (name=subject)
      ║     Var (name=_const_fb4ba132_uri, value=https://schema.org/identifier, anonymous)
      ║     Var (name=_const_d5c851ae_lit_e2eec718_0, value="27bccef5a7ef49dab2414f6eb6cd8ef4", anonymous)
      ╚══Join (JoinIterator) (resultSizeActual=14, totalTimeActual=37.6ms, selfTimeActual=4.2ms)
         ├──StatementPattern (costEstimate=1, resultSizeEstimate=7.6K, resultSizeActual=14, totalTimeActual=18.9ms, selfTimeActual=18.9ms)
         │     Var (name=subject)
         │     Var (name=_const_f5e5585a_uri, value=http://www.w3.org/1999/02/22-rdf-syntax-ns#type, anonymous)
         │     Var (name=_const_eaa085c9_uri, value=https://schema.org/CreativeWork, anonymous)
         └──StatementPattern (costEstimate=78, resultSizeEstimate=6.0K, resultSizeActual=14, totalTimeActual=14.6ms, selfTimeActual=14.6ms)
               Var (name=subject)
               Var (name=_const_880c9cef_uri, value=https://schema.org/version, anonymous)
               Var (name=version)

```

**Reply by naturzukunft** [reply]
There are other parallel queries running against the same database! One complex construct query. See [this thread](https://stackoverflow.com/questions/72250712/sparql-construct-query-performance-problem).
I still do some optimisation. currently the query above "getLatestVersion" seems to be the biggest problem.

|              Measurement Point             |   #  |  Average  |   Min   |     Max    |     Total     |
|--------------------------------------------|------|-----------|---------|------------|---------------|
| LoaRdfPubAdapter:createPublication         | 2267 | 1,595.033 | 424.587 | 10,894.563 | 3,615,940.617 |
|--------------------------------------------|------|-----------|---------|------------|---------------|
| LoaRdfPubAdapter:getConstructedPublication | 1874 |   217.148 | 150.446 |  1,481.410 |   406,934.767 |
|--------------------------------------------|------|-----------|---------|------------|---------------|
| LoaRdfPubAdapter:getLatestVersion          | 2267 | 1,060.500 | 147.258 | 10,179.626 | 2,404,154.529 |
|--------------------------------------------|------|-----------|---------|------------|---------------|
| LoaRdfPubAdapter:postPublication           | 2267 |   354.374 | 255.835 |  1,772.660 |   803,366.303 |
|--------------------------------------------|------|-----------|---------|------------|---------------|

**Reply by hmottestad** [reply]
Looks like the query is scaling ok. The actual data that the query is joining doesn't seem to be increasing and the join order doesn't look like it should matter.

**Reply by hmottestad** [reply]
Could it be that the JVM is selecting the SerialGC because of the limited heap size? https://stackoverflow.com/questions/70664562/criteria-for-default-garbage-collector-hotspot-jvm-11-17

**Reply by hmottestad** [reply]
There are also a few cache size settings you can try to increase. 

https://rdf4j.org/javadoc/3.7.7/org/eclipse/rdf4j/sail/nativerdf/NativeStore.html

![image](https://user-images.githubusercontent.com/797185/168855150-ceb096d6-4007-44c8-9c3b-98bf62be1bfa.jpeg)

**Reply by naturzukunft** [reply]
ok, i have continued debugging and testing. i have now outsourced the server part so that my laptop is relieved. in doing so, i have seen for the first time that the client builds up a lot of threads. i must take a closer look tomorrow. I have just seen that there are a lot of rdf4j-SharedHttpClientSessionManager-* threads. you will hear from me

**Comment by naturzukunft** [comment, marked as answer]
Shame on me. I found the problem. I was looking in completely the wrong place all along. The error was creating the SPARQLRepository on the client side. I created a new SPARQLRepository for each request. And i marked this code with a todo and a warning. But all this is of no use if you miss your hints.
Thanks again for your support! 
I'll now cleanup and move the whole stuff to the servers. 

Last test with running the client on my local machine:


|              Measurement Point             |   #   | Average |   Min   |   Max   |     Total     |
|--------------------------------------------|-------|---------|---------|---------|---------------|
| LoaRdfPubAdapter:createPublication         | 13792 | 274.594 | 173.967 | 948.103 | 3,787,206.738 |
|--------------------------------------------|-------|---------|---------|---------|---------------|
| LoaRdfPubAdapter:getConstructedPublication |  9092 |  89.409 |  65.313 | 700.371 |   812,909.246 |
|--------------------------------------------|-------|---------|---------|---------|---------------|
| LoaRdfPubAdapter:getLatestVersion          | 13792 |  78.690 |  53.920 | 694.373 | 1,085,293.251 |
|--------------------------------------------|-------|---------|---------|---------|---------------|
| LoaRdfPubAdapter:postPublication           | 13792 | 136.087 | 105.841 | 800.349 | 1,876,905.857 |
|--------------------------------------------|-------|---------|---------|---------|---------------|

### DURATION: 559 seconds for 5000 kvmEntries.
